### PR TITLE
Type annotations and fixes for backend/backends

### PR DESCRIPTION
**Pull request by dcbaker**
My goal was to add type annotations to backend/backends.py, as it seemed like a good starting point for looking into the backend. This ended up requiring a lot of fixes for the build module as well, and revealed several issues in already typed modules, like mesonlib. I've split each change into a single patch, as this is good practice, and makes bisecting easier for regressions.

### CODE REVIEW

File: mesonbuild/backend/backends.py [hunk 1]
```diff
@@ -135,13 +153,24 @@ def __init__(self, path: str, install_path: str, install_mode: 'FileMode', subpr
...
         self.exclude = exclude
 
 class ExecutableSerialisation:
-    def __init__(self, cmd_args, env: T.Optional[build.EnvironmentVariables] = None, exe_wrapper=None,
-                 workdir=None, extra_paths=None, capture=None, feed=None) -> None:
+
+    # XXX: should capture and feed default to False, instead of None?
+
+    def __init__(self, cmd_args: T.List[str],
+                env: T.Optional[build.EnvironmentVariables] = None,
```
Comment (mensinda) [inline mesonbuild/backend/backends.py:166]: bad indentation (and `cmd_args` could also go on a newline).

File: mesonbuild/backend/backends.py [hunk 2]
```diff
@@ -267,13 +299,17 @@ def get_compiler_options_for_target(self, target: build.BuildTarget) -> OptionOv
...
         if option_name in target.option_overrides_base:
             override = target.option_overrides_base[option_name]
-            return self.environment.coredata.validate_option_value(option_name, override)
-        return self.environment.coredata.get_option(option_name.evolve(subproject=target.subproject))
+            v = self.environment.coredata.validate_option_value(option_name, override)
+        else:
+            v = self.environment.coredata.get_option(option_name.evolve(subproject=target.subproject))
+        # We don't actually have wrapmode here to do an assert, so just do a
+        # cast, we know what's in coredata anyway.
+        return T.cast(T.Union[str, int, bool, 'WrapMode'], v)
```
Comment (mensinda) [inline mesonbuild/backend/backends.py:310]: This should be more a more noticeable todo to remove this cast once we have annotated coredata.

File: mesonbuild/backend/backends.py [hunk 3]
```diff
@@ -291,7 +327,7 @@ def get_build_dir_include_args(self, target, compiler, *, absolute_path=False):
                 curdir = '.'
         return compiler.get_include_args(curdir, False)
 
-    def get_target_filename_for_linking(self, target):
+    def get_target_filename_for_linking(self, target: T.Union[build.Target, build.CustomTargetIndex]) -> T.Optional[str]:
```
Comment (mensinda) [inline mesonbuild/backend/backends.py:330]: Is it really out of the question that not a `build.CustomTarget` can be passed here? There is an `isinstance` check for it below.

Same with the other functions below.

File: mesonbuild/backend/backends.py [hunk 4]
```diff
@@ -820,9 +905,9 @@ def generate_basic_compiler_args(self, target: build.BuildTarget, compiler: 'Com
         # command-line or default_options inside project().
         commands += compiler.get_option_compile_args(copt_proxy)
         # Add buildtype args: optimization level, debugging, etc.
-        commands += compiler.get_buildtype_args(self.get_option_for_target(OptionKey('buildtype'), target))
-        commands += compiler.get_optimization_args(self.get_option_for_target(OptionKey('optimization'), target))
-        commands += compiler.get_debug_args(self.get_option_for_target(OptionKey('debug'), target))
+        commands += compiler.get_buildtype_args(T.cast(str, self.get_option_for_target(OptionKey('buildtype'), target)))
+        commands += compiler.get_optimization_args(T.cast(str, self.get_option_for_target(OptionKey('optimization'), target)))
+        commands += compiler.get_debug_args(T.cast(bool, self.get_option_for_target(OptionKey('debug'), target)))
```
Comment (mensinda) [inline mesonbuild/backend/backends.py:910]: I would prefer asserts here. There is no reason that the types should ever accidentally change, but they do for some reason we will be glad that the asserts tell us early where the problem is.

File: mesonbuild/backend/backends.py [hunk 5]
```diff
@@ -910,46 +995,49 @@ def get_mingw_extra_paths(self, target):
                 paths.update(cc.get_library_dirs(self.environment))
         return list(paths)
 
-    def determine_windows_extra_paths(self, target: T.Union[build.BuildTarget, str], extra_bdeps):
-        '''On Windows there is no such thing as an rpath.
+    def determine_windows_extra_paths(
+            self, target: T.Union[build.BuildTarget, build.CustomTarget, programs.ExternalProgram, mesonlib.File, str],
```
Comment (mensinda) [inline mesonbuild/backend/backends.py:999]: I don't really know the code but from the current annotations this looks like a `build.CustomTargetIndex` or `build.GeneratedList` might also be passed here.

File: mesonbuild/backend/backends.py [hunk 6]
```diff
@@ -1337,7 +1427,8 @@ def create_install_data(self) -> InstallData:
                         self.environment.get_build_dir(),
                         self.environment.get_prefix(),
                         strip_bin,
-                        self.environment.coredata.get_option(OptionKey('install_umask')),
+                        # This will only be a string or number, but mypy doesn't know that.
+                        T.cast(T.Union[str, int], self.environment.coredata.get_option(OptionKey('install_umask'))),
```
Comment (mensinda) [inline mesonbuild/backend/backends.py:1431]: Again, an assert will be more code, but we really want this to blow up if something goes wrong.

**Review by mensinda — COMMENTED** [review]
I only got a few issues, but this LGTM otherwise.

File: mesonbuild/backend/backends.py (hunk 3 above)
Comment (dcbaker) [reply mesonbuild/backend/backends.py:330]: Custom target is a Target (though not a BuildTarget), but CustomTargetIndex is its own thing.

Comment (mensinda) [reply mesonbuild/backend/backends.py:330]: Oh, right. Makes sense.

File: mesonbuild/backend/backends.py (hunk 2 above)
Comment (dcbaker) [reply mesonbuild/backend/backends.py:310]: I don't think that we can properly annotate this, get_option returns  different type depending on the exact input, and I don't think there's a good way to annotate that, but I can add a better comment.

File: mesonbuild/backend/backends.py (hunk 5 above)
Comment (dcbaker) [reply mesonbuild/backend/backends.py:999]: I don't see any path where that could happen

**Review by mensinda — APPROVED** [review]
LGTM

**Pull request merged**